analyses/focal-cn-file-preparation/05-define-most-focal-cn-units.Rmd

```diff
@@ -0,0 +1,357 @@
+```
+
+## Set up
+
+### Libraries and functions
+
+```{r}
+library(tidyverse)
+```
+
+### Custom Function
+
+```{r}
+status_majority_caller <- function(status_df,
+                                   region_variable) {
+  # Given a data.frame with cytoband/gene level copy number status data,
+  # find the dominant status of the cytoband/gene region by calculating
+  # the percentages of the region that each call represents.
+  #
+  # Args:
+  #   status_df: data.frame with cytoband/gene level copy number status data
+  #   region_variable: string of the column name/region to calculate copy number
+  #                    status percentages for
+  # Return:
+  #   status_count: data.frame with percentage values for each unique status in
+  #                 each unique region (arm/cytoband/gene) and the dominant
+  #                 status for that region
+
+  region_sym <- rlang::sym(region_variable)
+
+  if (region_variable == "chromosome_arm") {
+    status_count <- status_df %>%
+      count(!!region_sym, dominant_status) %>%
+      # Spread the data -- each row represents a unique chromosome arm
+      spread(dominant_status, n) %>%
+      replace_na(list(
+        gain = 0,
+        loss = 0,
+        neutral = 0
+      )) %>%
+      group_by(!!region_sym) %>%
+      # The logic here is to define the arm status based on the majority of calls
+      # in the region -- if the number of calls for a specific status is
+      # responsible for more than 50% of the total calls in that region, then it is
+      # used to define the region's status
+      mutate(
+        percent_gain = gain / (gain + loss + neutral),
+        percent_loss = loss / (gain + loss + neutral),
+        percent_neutral = neutral / (gain + loss + neutral),
+        dominant_status = case_when(
+          percent_gain > 0.5 &
+            percent_gain > 1.5 * percent_loss ~ "gain",
+          percent_loss > 0.5 &
+            percent_loss > 1.5 * percent_gain ~ "loss",
+          percent_neutral > 0.5 &
+            percent_neutral > 1.5 * percent_loss &
+            percent_neutral > 1.5 * percent_gain ~ "neutral",
+          TRUE ~ "none"
+        )
+      )
+  } else {
+    status_count <- status_df %>%
+      count(!!region_sym, status) %>%
+      # Spread the data -- each row represents a unique cytoband/gene
+      spread(status, n) %>%
+      replace_na(list(
+        gain = 0,
+        loss = 0,
+        amplification = 0
+      )) %>%
+      group_by(!!region_sym) %>%
+      # The logic here is to define the arm status based on the majority of calls
+      # in the region -- if the number of calls for a specific status is
+      # responsible for more than 50% of the total calls in that region, then it is
+      # used to define the region's status
+      mutate(
+        percent_gain = gain / (gain + loss + amplification),
+        percent_loss = loss / (gain + loss + amplification),
+        percent_amplification = amplification / (gain + loss + amplification),
```

Comment:
Looks like the only difference here is that its reporting `amplification` instead of `neutral`?

Reply:
I think we can still shorten up this section.

Reply:
Let me know if you wanna pair code on this. I have some ideas on this, but might be easiest to pair up on it.

Reply:
Sure we pair on this. Would you prefer that I address your other comments and push those changes first? Or woyld you rather pair on it sooner?

Reply:
Go ahead and address the comments I have and then maybe we should schedule a time to try to rework this main function.